src/lib/format/l20n/parser.js

```diff
@@ -5,201 +5,134 @@ import { L10nError } from '../../errors';
-      this.getValue(false, undefined, index)
-    ];
-  },
-
-  getHash: function() {
-    ++this._index;
-    this.getWS();
-    let hash = {};
-    while (true) {
-      const hi = this.getKVP();
-      hash[hi[0]] = hi[1];
-      this.getWS();
-
-      const comma = this._source.charAt(this._index) === ',';
-      if (comma) {
-        ++this._index;
-        this.getWS();
-      }
-      if (this._source.charAt(this._index) === '}') {
+      const id = this.getIdentifier();
+      if (this._source[this._index] === '[') {
         ++this._index;
-        break;
-      }
-      if (!comma) {
-        throw this.error('Expected "}"');
+        return this.getEntity(id, this.getItemList(this.getExpression, ']'));
       }
+      return this.getEntity(id);
     }
-    return hash;
-  },
 
-  unescapeString: function(str, opchar) {
-    function replace(match, p1) {
-      switch (p1) {
-        case '\\':
-          return '\\';
-        case '{{':
-          return '{{';
-        case opchar:
-          return opchar;
-        default:
-          if (p1.length === 5 && p1.charAt(0) === 'u') {
-            return String.fromCharCode(parseInt(p1.substr(1), 16));
-          }
-          throw this.error('Illegal unescape sequence');
-      }
+    if (this._source.startsWith('/*', this._index)) {
+      return this.getComment();
     }
-    return str.replace(this._patterns.unesc, replace.bind(this));
+
+    throw this.error('Invalid entry');
   },
 
-  getString: function(opchar) {
-    let opcharPos = this._source.indexOf(opchar, this._index + 1);
+  getEntity: function(id, index) {
+    if (!this.getRequiredWS()) {
+      throw this.error('Expected white space');
+    }
 
-    outer:
-    while (opcharPos !== -1) {
-      let backtrack = opcharPos - 1;
-      // 92 === '\'
-      while (this._source.charCodeAt(backtrack) === 92) {
-        if (this._source.charCodeAt(backtrack - 1) === 92) {
-          backtrack -= 2;
-        } else {
-          opcharPos = this._source.indexOf(opchar, opcharPos + 1);
-          continue outer;
+    const ch = this._source[this._index];
+    const value = this.getValue(ch, index === undefined);
+    let attrs;
+
+    if (value === undefined) {
+      if (ch === '>') {
+        throw this.error('Expected ">"');
+      }
+      attrs = this.getAttributes();
+    } else {
+      const ws1 = this.getRequiredWS();
+      if (this._source[this._index] !== '>') {
+        if (!ws1) {
+          throw this.error('Expected ">"');
         }
+        attrs = this.getAttributes();
       }
-      break;
-    }
-
-    if (opcharPos === -1) {
-      throw this.error('Unclosed string literal');
     }
 
-    let buf = this._source.slice(this._index + 1, opcharPos);
-
-    this._index = opcharPos + 1;
+    // skip '>'
+    ++this._index;
 
-    if (!this.simpleMode && buf.indexOf('\\') !== -1) {
-      buf = this.unescapeString(buf, opchar);
+    if (id in this.entries) {
+      throw this.error('Duplicate entry ID "' + id);
```

Comment:
I don't think I can use the same message because parser doesn't know which language it parses (and I think this separation is a good thing). What we could do is differentiate types of errors in the parser, and then in env when we catch duplicateentry error report it properly. What do you think?